client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -144,27 +145,38 @@ private static <T> Iterator<T> flatten(Iterator<Iterable<T>> iterator, long maxR
         return stream.iterator();
     }
 
-    private static class AsyncIterator<T>
+    @VisibleForTesting
+    static class AsyncIterator<T>
             extends AbstractIterator<T>
     {
         private static final int MAX_QUEUED_ROWS = 50_000;
         private static final ExecutorService executorService = newCachedThreadPool(
                 new ThreadFactoryBuilder().setNameFormat("Trino JDBC worker-%s").setDaemon(true).build());
 
         private final StatementClient client;
-        private final BlockingQueue<T> rowQueue = new ArrayBlockingQueue<>(MAX_QUEUED_ROWS);
+        private final BlockingQueue<T> rowQueue;
         // Semaphore to indicate that some data is ready.
         // Each permit represents a row of data (or that the underlying iterator is exhausted).
         private final Semaphore semaphore = new Semaphore(0);
-        private final CompletableFuture<Void> future;
+        private final Future<?> future;
+        private volatile boolean cancelled;
+        private volatile boolean finished;
```

Comment:
What's the difference between finished and cancelled? why do we need two fields?

Reply:
- `cancelled` is used for process control. When `cancel()` is invoked, `cancelled` is set to `true` so that the iteration can be terminated
- and `finished` is a state used only for test to check whether the underlying work is done or terminated. 
 
The issue we want to solve here is that threads may leak even though we cancel or close the ResultSet. We should make sure that threads can finally complete the task in the `run` and should not get blocked in the middle.